lib/active_merchant/billing/gateways/rapyd.rb

```diff
@@ -0,0 +1,239 @@
+        post[:payment_method][:type] = type
+
+        if type.include?('card')
+          add_creditcard(post, payment)
+        elsif type.include?('bank')
+          add_ach(post, payment, options)
+        end
+      end
+
+      def add_creditcard(post, payment)
+        post[:payment_method][:fields][:number] = payment.number
+        post[:payment_method][:fields][:expiration_month] = payment.month.to_s
+        post[:payment_method][:fields][:expiration_year] = payment.year.to_s
+        post[:payment_method][:fields][:cvv] = payment.verification_value.to_s
+        post[:payment_method][:fields][:name] = "#{payment.first_name} #{payment.last_name}"
+      end
+
+      def add_ach(post, payment, options)
+        post[:payment_method][:fields][:proof_of_authorization] = options[:proof_of_authorization]
+        post[:payment_method][:fields][:first_name] = payment.first_name if payment.first_name
+        post[:payment_method][:fields][:last_name] = payment.last_name if payment.last_name
+        post[:payment_method][:fields][:routing_number] = payment.routing_number
+        post[:payment_method][:fields][:account_number] = payment.account_number
+        post[:payment_method][:fields][:payment_purpose] = options[:payment_purpose] if options[:payment_purpose]
+      end
+
+      def parse(body)
+        return {} if body.empty? || body.nil?
+
+        JSON.parse(body)
+      end
+
+      def commit(method, action, parameters)
+        url = (test? ? test_url : live_url) + action.to_s
+        rel_path = "#{method}/v1/#{action}"
+        response = api_request(method, url, rel_path, parameters)
+
+        Response.new(
+          success_from(response),
+          message_from(response),
+          response,
+          authorization: authorization_from(response),
+          avs_result: avs_result(response),
+          cvv_result: cvv_result(response),
+          test: test?,
+          error_code: error_code_from(response)
+        )
+      end
+
+      def api_request(method, url, rel_path, params)
+        params == {} ? body = '' : body = params.to_json
+        parse(ssl_request(method, url, body, headers(rel_path, body)))
+      end
+
+      def headers(rel_path, payload)
+        salt = SecureRandom.base64(12)
+        timestamp = Time.new.to_i.to_s
+        {
+          'Content-Type' => 'application/json',
+          'access_key' => @options[:access_key],
+          'salt' => salt,
+          'timestamp' => timestamp,
+          'signature' => generate_hmac(rel_path, salt, timestamp, payload)
+        }
+      end
+
+      def generate_hmac(rel_path, salt, timestamp, payload)
+        signature = "#{rel_path}#{salt}#{timestamp}#{@options[:access_key]}#{@options[:secret_key]}#{payload}"
+        hash = Base64.urlsafe_encode64(OpenSSL::HMAC.hexdigest('sha256', @options[:secret_key], signature))
+        hash
+      end
+
+      def avs_result(response)
+        case resp = response['data']
+        when !resp.nil?
```

Comment:
I will admit being confused by `!resp.nil?`

Could this method be rewritten like:
```
def avs_result(response)
  return nil unless (code = response.dig('data', 'payment_method_data', 'acs_check'))

  AVSResult.new(code: code)
end
```
...or maybe I am misunderstanding what is happening here?

Reply:
This is much cleaner than my solution, thank you! Revised!